---
title: "mega-non-model-wgs-snakeflow"
output: github_document
---

## Quick install and run

If you would like to put this on your system and test it running on a single node
(more later about using SLURM for deployment across multiple nodes) you have to
clone this repository and then download the pseudo-genome used for
the included test data set (in `.test`).

You must have Snakemake (version > 6.0) in the active environment.

In short, here are the steps to install and run the `.test`.

```sh

```

## What the user must do and values to be set, etc

* Choose an Illuminaclip adapter fasta (in config)



## Assumptions

* Paired end


## Things fixed or added relative to JK's snakemake workflow

* fastqc on both reads
* don't bother with single end
* add adapters so illumina clip can work
* benchmark each rule
* use genomicsDBimport
* allow for merging of lots of small scaffolds into genomicsDB
